app/prop29.go

```diff
@@ -338,82 +367,88 @@ func verifyBitcoinSignature(sig, msg, addr string) {
 
 	address, err := btcutil.NewAddressPubKey(serializedPK, &chaincfg.MainNetParams)
 	if err != nil {
-		log.Fatal("Address recovery failed")
-
+		return fmt.Errorf("address recovery from pubkey failed: %w", err)
 	}
 
 	if address.EncodeAddress() != addr {
-		log.Fatal("invalid signature")
+		return fmt.Errorf("address mismatch, expected %s, got %s", addr, address.EncodeAddress())
 	}
+
+	return nil
 }
 
 func Prop29Migration(authGenesis *authtypes.GenesisState, bankGenesis *banktypes.GenesisState, distrGenesis *distr.GenesisState) (authtypes.GenesisState, banktypes.GenesisState, distr.GenesisState) {
-
 	fundRecovery := validateFundRecovery()
 
 	recoveryAccounting := sdk.NewInt64Coin("uatom", 0)
-
 	emptyCoins := []sdk.Coin{}
-
 	distModuleAccount := authtypes.NewModuleAddress(distr.ModuleName)
 
-	//Set All Source Addresses to Zero and accumulate the total funds being moved
+	// zero out all source addresses balances and accumulate the total funds being moved
 	for i, balance := range bankGenesis.Balances {
 		_, isSourceAddress := fundRecovery.IsSourceAddress(balance.Address)
-		if isSourceAddress {
-			if len(balance.Coins) > 1 {
-				log.Fatal("Expected all balances to contain only 1 denom during the migration")
-			}
-			// Accumulate all the coins removed from the balances
-			recoveryAccounting = recoveryAccounting.Add(balance.Coins[0])
-			// Zero out the Balance
-			bankGenesis.Balances[i].Coins = emptyCoins
+		if !isSourceAddress {
+			continue
 		}
 
+		if len(balance.Coins) > 1 {
+			log.Fatal("expected all balances to contain only 1 denom during the migration")
+		}
+		// accumulate all the coins removed from the balances into the pool
+		recoveryAccounting = recoveryAccounting.Add(balance.Coins[0])
+		// Empty the source address balance
+		bankGenesis.Balances[i].Coins = emptyCoins
 	}
 
+	// migrate the balances to the the destination addresses
 	for i, balance := range bankGenesis.Balances {
 		index, isDestAddress := fundRecovery.IsDestAddress(balance.Address)
-		if isDestAddress {
-			recoveryAccounting = recoveryAccounting.Sub(fundRecovery[index].destBalance)
-			bankGenesis.Balances[i].Coins = bankGenesis.Balances[i].Coins.Add(fundRecovery[index].destBalance)
-			fundRecovery[index].destBalance = sdk.NewInt64Coin("uatom", 0)
+		if !isDestAddress {
+			continue
 		}
+
+		// transfer coins from the atom recovery pool to the dest address
+		recoveryAccounting = recoveryAccounting.Sub(fundRecovery[index].destBalance)
+		bankGenesis.Balances[i].Coins = bankGenesis.Balances[i].Coins.Add(fundRecovery[index].destBalance)
+		fundRecovery[index].destBalance = sdk.NewInt64Coin("uatom", 0)
 	}
+
+	// add the balances to the bank genesis
 	bankGenesis.Balances = append(bankGenesis.Balances, fundRecovery.GetRemainingBalances()...)
```

Comment:
My only concern here is that a balance uniqueness is not validated on the bank genesis validation